### ISSUE TITLE
Anchors with too small strides matched to large target boxes

### ISSUE THREAD

**Issue by Adamusen** [WongKinYiu/YOLO#124]
**Issue Description**

The way box coordinate predictions work in yolo, is that the regression problem is solved using classification instead. There are a set of anchor points in multiple grids (in the standard configuration three grids with **strides=[8, 16, 32]**), and the box predictor head makes a classification between a given amount of classes, provided by **reg_max** (default 16) for each box coordinate (xmin, ymin, xmax, ymax). These classes are defined in a way, in which they predict, by _how many amount of times stride_ a given box coordinate value differs from the position of the anchor itself. So if an anchor with stride 8 for example predicts class 4 out of 16 for its xmax value. It means the final bounding box xmax value will be: `x_max = anchor_x + (4-1) * 8`, where 4-1, because the first class out of 16 stands for zero difference. This has two implications:
 - Anchors are only able to predict bounding boxes, if the anchor is inside the target box, because only positive difference values can be predicted for a_x - x_min, a_y - y_min, x_max - a_x, y_max - a_y, where a_x and a_y are the anchor's position.
 - There is a maximum amount of difference that can be predicted for each value, limiting the maximum size of the bounding box. The maximum difference in pixels is: `max_diff_pixels = (reg_max - 1) * stride`. This means, that an anchor with stride 8 for example, is only capable of predicting an xmax value, which is at most 120 pixels away from its base, or a bounding box, that is 240x240 pixels at most, assumed the anchor base sits perfectly in the middle of the target.

In the loss function, target bounding boxes are assigned a set of best suited anchors (for which the maximum amount is loss/matcher/topk in train.yaml), which are trained to predict them. The first constraint from above is satisfied as the first step within the BoxMatcher within yolo/utils/bounding_box_utils.py, in which at Line 267 the grid_mask is calculated, that tells which anchors lie within which target (ground truth) bounding box, so that only these anchors might get assigned to said target boxes.

For the second constraint, however, there exists no kind of filtering within the implementation, which leads to the fact, that anchors with smaller strides are able to get assigned to larger target boxes, which they are unable to predict. To test this, I built the following check within yolo/tools/loss_functions.py inside the DFLoss classes forward function, starting at Line 47:

```
    def forward(
        self, predicts_anc: Tensor, targets_bbox: Tensor, valid_masks: Tensor, box_norm: Tensor, cls_norm: Tensor
    ) -> Any:
        valid_bbox = valid_masks[..., None].expand(-1, -1, 4)
        bbox_lt, bbox_rb = targets_bbox.chunk(2, -1)
        # targets_dist = torch.cat(((self.anchors_norm - bbox_lt), (bbox_rb - self.anchors_norm)), -1).clamp(
        #     0, self.reg_max - 1.01
        # )
        targets_dist = torch.cat(((self.anchors_norm - bbox_lt), (bbox_rb - self.anchors_norm)), -1)

        picked_targets_noclamp = targets_dist[valid_bbox].view(-1)
        picked_targets_clamp0 = targets_dist[valid_bbox].view(-1).clamp(0)
        picked_targets = targets_dist[valid_bbox].view(-1).clamp(0, self.reg_max - 1.01)
        clamped_num = torch.sum(picked_targets_noclamp != picked_targets)
        print(f"{int(torch.sum(picked_targets_noclamp != picked_targets_clamp0))}",
              f"{clamped_num}/{picked_targets_noclamp.size(0)}",
              f"{int(clamped_num) / picked_targets_noclamp.size(0) * 100:.2f}%")
```

This part of the code is responsible to train the above mentioned classification part of the box prediction head. For this, it calculates said distance between the anchor and the various box coordinates first, and clamps these distances between 0 and reg_max - 1.01, since this is a range limit each anchor is able to predict. What my short test does here, is moves the clamping after the valid anchors are picked (the ones that actually got assigned) and checks, how many times did the clamp() call actually clipped the values. This is what I got:

```
0 69/880 7.84%
0 55/880 6.25%
0 54/960 5.62%
0 53/964 5.50%
0 5/1148 0.44%
0 0/1148 0.00%
0 168/752 22.34%
0 171/760 22.50%
0 106/1744 6.08%
0 44/1744 2.52%
0 63/480 13.12%
0 52/480 10.83%
0 17/400 4.25%
0 14/400 3.50%
0 40/976 4.10%
0 21/948 2.22%
0 67/2424 2.76%
0 8/2420 0.33%
0 47/1240 3.79%
0 38/1244 3.05%
0 51/1768 2.88%
0 39/1788 2.18%
0 134/1148 11.67%
0 126/1140 11.05%
0 20/1704 1.17%
```

These values mean the following:
- How many values were negativ and got clamped to zero from the left (never occurs, which is good, it means the first constraint is satisfied)
- How many of all the assigned anchors got clipped from the right (the stride times distance was above reg_max - 1.01)
- The same as above in percentiles

What this means is, a significant amount of anchors are being trained to predict targets, which they are unable to do so, because they are larger than what the anchor can predict. These results are logically worse the higher the resolution one trains with, I got these values with a test training with a relatively high resolution of 1920x1280.

**Proposed solution**

Create an additional mask in addition to the "grid_mask" which excludes all anchors from each target box, which they are unable to predict, because any of the target coordinates is more distant to the anchor's base as `stride * (reg_max - 1)` would allow to predict.

This on its own proposes a new issue, that is, if a ground truth bounding box is so large, that no anchor is fit to predict it, it will simply be ignored during training, and the network will never learn to predict it at all. As a compromise, I propose to only satisfy the second constraint for anchors with lower strides (in the default case 8 and 16) and let the anchors with the highest strides (default 32) to be still matched to targets they can't perfectly predict, as a somewhat smaller prediction is still better than nothing. This is only a compromise to avoid this issue, if ground truth boxes are too large because of high resolution for example, the user should either increase the reg_max value or create a configuration with an additional pyramid layer with an even higher stride (64) to handle these targets anyway, however, such changes require a fairly high understanding of how yolo box predictions work from the user.

I will implement this solution and test it out myself, as well as create a pull request with it, however, since this is not something that's part of the "stable" yolo implementation either, I fully understand if it does not eventually get merged.

What do you think about this problem @henrytsui000 ?

**Comment by henrytsui000** [comment]
Hi,

I completely agree with your idea! Additionally, I’d like to ask: how much smaller is the grid size compared to the target box? I think this could help predict how effective the approach will be—it’s bound to be helpful to some degree.

Regarding the first constraint, I consider having grid points outside of the box sometimes to be acceptable (see Figure 1). Sometimes, it can even provide a halfway result. However, this is just my hypothesis, and I plan to optimize this after the training process becomes stable.

![image](https://github.com/user-attachments/assets/cb853fb9-b6cf-4e16-a3cc-1b4fe91aaf74)

In summary, if your implementation can improve the coverage, I will definitely merge it!

Best regards,
Henry Tsui

**Comment by Adamusen** [comment]
Hi @henrytsui000,

**First point**

To answer your first question, I extended the print statement in my test with the following Line:
`f"{torch.mean((picked_targets_noclamp-picked_targets)[picked_targets_noclamp != picked_targets]):2f}"`,
and the new output looks like this:

```
0 43/1080 3.98% 2.435414
0 15/1080 1.39% 1.522708
0 95/1276 7.45% 5.376155
0 58/1280 4.53% 3.831054
0 67/920 7.28% 4.091290
0 67/920 7.28% 3.941620
0 142/1224 11.60% 6.416493
0 125/1228 10.18% 6.333450
0 68/1180 5.76% 3.729699
0 54/1172 4.61% 4.730016
0 52/3124 1.66% 2.512861
0 29/3072 0.94% 2.737645
0 73/1116 6.54% 1.455466
0 69/1104 6.25% 1.560206
0 63/680 9.26% 6.435242
0 65/680 9.56% 5.875626
0 107/1480 7.23% 6.407396
0 101/1480 6.82% 6.447715
0 46/1236 3.72% 2.460328
0 34/1240 2.74% 2.217988
```

which means, whenever the values are clipped, they are clipped by ~2-6, most likely from anchors with stride 8, so on avarage a lot of anchors are trained to learn a box that is 16-48 (on average, sometimes even more!) pixels smaller in at least one of the four directions as the target.

**Second point**

Regarding your second point about the first constraint. Because of that, yes, there is a smallest possible box the network is able to target as well currently (which is 8x8 at the moment), since if a target is smaller, it might not be assigned any anchor if it does not lie on any anchor point by sheer accident as you demonstrated on your Figure 1.

This problem can be mitigated by your proposal, but it would still not allow the target to be predicted correctly, only to be fitted with a slightly larger box. For this problem, I came up with an even better solution in my opinion, which is slightly more complex to implement, but far from being too bad. It is the following:

Lets add an additional parameter in the model.yaml file's anchor configuration, which could be called "reg_start" (next to reg_max and strides) for example. This parameter, if not contained, would be assumed to be zero for compatibility with already trained checkpoints, but in the future could be configured to be -1.

What reg_start would do, is set the starting value of the arange within the Detection head's Anchor2Vec module currently at Line 175 in yolo/model/module.py:

`reverse_reg = torch.arange(reg_max, dtype=torch.float32).view(1, reg_max, 1, 1, 1)` ->
`reverse_reg = torch.arange(reg_start, reg_max, dtype=torch.float32).view(1, reg_max, 1, 1, 1)`,

with this modification and a reg_start value of -1, the predicted classes would be [-1, 0, 1..., 14] instead of [0, 1, 2..., 15]. This would allow each anchor to predict xmin, ymin, xmax, ymax values up to one stride distance in the negative direction as well!

This change of course would require modifications in the BoxMatcher, DFLoss and PostProcessing to also be taken in to consideration to calculate the values properly, but it would enable the BoxMatcher to match anchors up to one stride distance outside the target box, which would not only eliminate the smallest predictable target box problem, but would guarantee, that at least 4 anchors can be matched for even the smallest of targets and even able to predict it properly!

What do you think about this idea?

**Comment by Adamusen** [comment]
Hi @henrytsui000,

I implemented the first part of the proposed solution, in which I mask out anchors from targets in BoxMatcher/get_valid_matrix if they can't predict it considering both of the discussed criteria (1, not inside; 2, box too large).

On top of this, I implemented a compromise solution for the problem, if a target can't be matched with any anchor, because it is either too small or too large to be perfectly predicted, in which I match one best suited anchor to such a target, even if it can't predict it perfectly, as a somewhat scewed box prediction is still better than being ignored completely. I called this function ensure_one_anchor, which follows the topk selection.

Testing the amount of clipping with the impelementation, I get the following:
```
0 0/1328 0.00% nan
0 0/1320 0.00% nan
0 0/1696 0.00% nan
0 0/1692 0.00% nan
0 1/964 0.10% 1.094376
0 2/964 0.21% 0.279687
0 0/1300 0.00% nan
0 0/1296 0.00% nan
0 0/1336 0.00% nan
0 0/1340 0.00% nan
0 0/800 0.00% nan
0 0/800 0.00% nan
0 0/800 0.00% nan
0 0/800 0.00% nan
0 2/1084 0.18% 4.997499
0 2/1084 0.18% 4.997499
0 1/1200 0.08% 3.284399
0 2/1184 0.17% 1.203600
```

, it is reduced to almost zero! The only time some clipping happens, is when my "ensure_one_anchor" matches an anchor to a target, which is too large or too small. If I comment it out, I always get zero clipping, but the fact, that with it there is sometimes some shows, that the function works as intended :)

I already trained a couple of epochs with it starting from an already trained checkpoint, and the network is still doing good, so the solution looks to work properly. I let it run for a little bit more and if it looks good I'll create a pull request.
